Port Competition Results to React

saranshgrover opened the pull request:
Closes #5455 

This is based off the redesign PR and creates a new React Component called `CompetitionResults` along with API endpoints for accessing them that @viroulep added. Here's a picture of what it looks like:
![image](https://user-images.githubusercontent.com/24918614/88148881-42f04c00-cbcd-11ea-9d53-8f10cfb983fa.png)

I had a couple questions that I was unsure about:

1. I'm not sure if we should still use the parameters now that the data is loaded dynamically with react. I currently wrote it such that if there is a parameter specified, the component adheres to it but when switching between event results, it doesn't update the URL params according.

2. I simply created a navigation component specific to this, since I didn't see a react specific component, but would it make more sense to have a more generic event navigation component similar to other parts of the site to stay consistent?

3. The Solves list looks pretty janky on mobile, I think it might be good to hide it, but that would restrict mobile users from viewing solves. We could do what WCA Live does and implement a modal if the user is on mobile to show solves, or maybe just a tooltip over the average cell?

viroulep commented:
Thanks for the draft @saranshgrover!

For 1: yes, I think we should keep (and update) an url parameter.
The rationale is that we want users to be able to link to competition's results for any event, and have a displayed state which is consistent with the url.
To get or set url params you can checkout [these](https://github.com/thewca/worldcubeassociation.org/blob/master/WcaOnRails/app/webpacker/javascript/wca/utils.js#L16) functions in utils!

For 2: I think the approach here would be to use existing components from FUI, I'm thinking a [text menu](https://react.semantic-ui.com/collections/menu/#types-text) where we display event icons instead of text.
I think an "events navigation" is something we're likely to re-use, so feel free to create a component for that!

For 3: don't tables stack on mobile? As far as I can see in [here](https://react.semantic-ui.com/collections/table/#types-pagination) it should.
In any case, I like your suggestion of hiding the individual times when we don't have the space for it, and using a modal to show them.
 
Some other UI suggestions for the tables:
  - could you make the table `striped`? It would help keeping track of each line :)
  - I think we should have some kind of fixed width for cells, as currently all rounds don't have the same cell width.

viroulep commented on WcaOnRails/app/webpacker/javascript/competition_results/index.js:
```diff
@@ -0,0 +1,112 @@
+          <Table.Row key={result.id}>
+            <Table.Cell>{result.pos}</Table.Cell>
+            <Table.Cell>
+              <a href={`/persons/${result.wca_id}`}>{`${result.name}`}</a>
```
I've started a "routes" module [here](https://github.com/thewca/worldcubeassociation.org/pull/5340/files#diff-40e8a9d588d2ad70187890336dd4bb09), would you mind adding the various website urls you used in this component there?

The rationale behind this is:
  - to group the routes used in js in a common place
  - but most importantly to make sure to rely on rails to provide us with the appropriate path, in case some day it's slightly altered (in the specific case of `persons` I highly doubt it will ever happens, but you get the point).

To give you an example, a helper for that route could look like:
```js
export const personUrl = (wcaId) => {
  return `<%= URI.decode(Rails.application.routes.url_helpers.person_path("${wcaId}")) %>`;
};
```
(please note I did *not* run the code, apologies if it's slightly off!)

saranshgrover replied:
@viroulep  Would you like me to add the API paths like this as well? If so, could you guide me on where I can access them through Rails? I'm unable to figure that out.

viroulep replied:
Yes and yes, will try to give you an example when I can!

viroulep replied:
If you want to get a link for this for instance:
```
  const { loading, error, data } = useLoadedData(
    `/api/v0/competitions/${competitionId}/results/${eventId}`
  );
```
You would probably want to expose a route like that:
```js
export const competitionResultsApiUrl = (id, eventId) => `<%= URI.decode(Rails.application.routes.url_helpers.api_v0_competition_results_path("${id}", "${eventId}")) %>`;
```

I tested this locally and it didn't work because somehow adding `(/:eventId)` to the route prevent rails from generating its name automatically :s
For this to work you will have to change this route:
```ruby
        get '/results(/:event_id)' => 'competitions#results'
```
to
```ruby
        get '/results(/:event_id)' => 'competitions#results', as: :results
```

You can get a list of all routes and their generated name by running `bin/rails routes`; alternatively you can also start a rails console with `bin/rails c` and take a look at `Rails.application.routes.url_helpers.methods` which includes all routes exposed by rails.

saranshgrover commented:
I've added support to update URL Parms, as well as extracted navigation to an `EventNavigation` component. Also changed the styles that you requested. For some reason, the Table styles weren't being applied initially - they do seem to work now and the tables do stack, so I don't think a modal is necessary.

UnsolvedCypher commented:
Is this ready to be merged?

saranshgrover commented:
This PR should be good now 😃  @jonatanklosko would you mind looking over it?

jonatanklosko commented on WcaOnRails/app/webpacker/javascript/event_navigation/index.js:
```diff
@@ -0,0 +1,31 @@
+
+const EventNavigation = ({ selected, eventIds, onSelect }) => (
+  <Menu text>
+    {eventIds.map((event, index) => (
```
Nit: please name this `eventId` instead, as it may cause some confusion. For instance there's `eventIds[index]` below, whereas it could just be `eventId`.

jonatanklosko commented on WcaOnRails/app/webpacker/javascript/competition_results/index.js:
```diff
@@ -0,0 +1,96 @@
+import React, { useState, useEffect } from 'react';
+import { Table } from 'semantic-ui-react';
+import useLoadedData from '../requests/loadable';
```
Could you rename `loadable` to `useLoadedData`? :>

jonatanklosko commented on WcaOnRails/app/webpacker/javascript/competition_results/index.js:
```diff
@@ -0,0 +1,96 @@
+      setUrlParams({ event: selectedEvent });
+    }
+  }, [selectedEvent]);
+  if (loading || selectedEvent === undefined) return <Loading />;
```
Nit: I'd use `useState(null)` and compare to `null` here as it's a bit cleaner (subjective I guess).

Also an empty line between `useEffects`, `return`, etc. helps readability a bit. Usually a multiline expression deserves a space around.

jonatanklosko commented on WcaOnRails/app/webpacker/javascript/wca-live/attempts.js:
```diff
@@ -0,0 +1,154 @@
+export const trimTrailingZeros = (array) => {
```
This file has been taken from `wca-live`, but we only need the result formatting function. Also sice then `wca-live` has been re-written. Here's the relevant part:

```js
export const SKIPPED_VALUE = 0;
export const DNF_VALUE = -1;
export const DNS_VALUE = -2;

/**
 * Converts the given attempt result to a human-friendly string.
 *
 * @example
 * formatAttemptResult(-1, '333'); // => 'DNF'
 * formatAttemptResult(6111, '333'); // => '1:01.11'
 * formatAttemptResult(900348002, '333mbf'); // => '11/13 58:00'
 */
export function formatAttemptResult(attemptResult, eventId) {
  if (attemptResult === SKIPPED_VALUE) return '';
  if (attemptResult === DNF_VALUE) return 'DNF';
  if (attemptResult === DNS_VALUE) return 'DNS';
  if (eventId === '333mbf') return formatMbldAttemptResult(attemptResult);
  if (eventId === '333fm') return formatFmAttemptResult(attemptResult);
  return centisecondsToClockFormat(attemptResult);
}

/**
 * Converts centiseconds to a human-friendly string.
 */
function centisecondsToClockFormat(centiseconds) {
  if (!Number.isFinite(centiseconds)) {
    throw new Error(
      `Invalid centiseconds, expected positive number, got ${centiseconds}.`
    );
  }
  return new Date(centiseconds * 10)
    .toISOString()
    .substr(11, 11)
    .replace(/^[0:]*(?!\.)/g, '');
}

function formatMbldAttemptResult(attemptResult) {
  const { solved, attempted, centiseconds } = decodeMbldAttemptResult(
    attemptResult
  );
  const clockFormat = centisecondsToClockFormat(centiseconds);
  const shortClockFormat = clockFormat.replace(/\.00$/, '');
  return `${solved}/${attempted} ${shortClockFormat}`;
}

/**
 * Returns an object representation of the given MBLD attempt result.
 *
 * @example
 * decodeMbldAttemptResult(900348002); // => { solved: 11, attempted: 13, centiseconds: 348000 }
 */
function decodeMbldAttemptResult(value) {
  if (value <= 0) return { solved: 0, attempted: 0, centiseconds: value };
  const missed = value % 100;
  const seconds = Math.floor(value / 100) % 1e5;
  const points = 99 - (Math.floor(value / 1e7) % 100);
  const solved = points + missed;
  const attempted = solved + missed;
  const centiseconds = seconds === 99999 ? null : seconds * 100;
  return { solved, attempted, centiseconds };
}

function formatFmAttemptResult(attemptResult) {
  /* Note: FM singles are stored as the number of moves (e.g. 25),
     while averages are stored with 2 decimal places (e.g. 2533 for an average of 25.33 moves). */
  const isAverage = attemptResult >= 1000;
  return isAverage
    ? (attemptResult / 100).toFixed(2)
    : attemptResult.toString();
}
```

jonatanklosko reviewed (commented):
Great you managed to resolve the commit history. This looks great, a few comments from me =)

jonatanklosko reviewed (approved):
LGTM, thanks!

jonatanklosko commented on WcaOnRails/app/controllers/api/v0/competitions_controller.rb:
```diff
@@ -29,7 +29,34 @@ def schedule
   def results
     competition = competition_from_params
-    render json: competition.results
+    render json: competition
```
Nit: this should actually stay as `competition.results` (that's why Travis is failing by the way ^^).

jonatanklosko commented on WcaOnRails/app/controllers/api/v0/competitions_controller.rb:
```diff
@@ -29,7 +29,34 @@ def schedule
+  def event_results
+    competition = competition_from_params
+    event = Event.c_find(params[:event_id])
+    if event
```
Ah, and we no longer need this conditional block. We have a separate path for `/results` and `/results/:event_id`, so here we can just do `Event.c_find!(params[:event_id])` to throw exception for invalid id.

DanielEgdal commented:
I think there is a mistake in the changes. When you go to a competition with results, like this: https://www.worldcubeassociation.org/competitions/ShenzhenGoodAfternoon2020/results/all?event=333

Then when you click on a person it will not direct to their profile correctly. Example with the winner of the competition:
https://www.worldcubeassociation.org/persons/wca_id:%202017XURU04
The link should not include this part "wca_id:%20". I believe this makes sense why the error occurs when you look at a line 1 of the file "WcaOnRails/app/webpacker/javascript/requests/routes.js.erb". Not fully certain though, so I am just highlighting it here.

saranshgrover commented:
Thanks for noticing! sent a fix in #5906

SAuroux commented:
@saranshgrover Uhm...it looks like this PR erased the `event=all` option used by WRT to look through all results during the posting procedure. Could this please become available again?

saranshgrover commented:
@SAuroux could you clarify what `event=all` did? It is still possible to look through all results, just not all at once. Would you like `event=all` to show all results at once?

Jambrose777 commented:
> @SAuroux could you clarify what `event=all` did? It is still possible to look through all results, just not all at once. Would you like `event=all` to show all results at once?

Yes that’s what event=all did. WRT pretty much needs this when posting results so that we can quickly look through all the results at once (for the bigger comps having to do different searches / eyeball scans on 18 different pages isn’t feasible for each of our members to do)

SAuroux commented:
@saranshgrover just to make sure: I hope you are not waiting for my answer? @Jambrose777 already summarized this perfectly on behalf of WRT, so I did not answer so far.

saranshgrover commented:
Yes I'm not waiting, will get to this asap!

saranshgrover commented:
#5947 has the fix

SAuroux commented:
I just tested it and everything looks good to me, thank you!
